In [2]:
import pandas as pd
import numpy as np

Configura o pandas para não ter limite de linhas

In [3]:
pd.set_option("max_rows", None)

# Teste 2

In [4]:
# Teste 2
raw_test2_data = pd.read_csv("teste2.csv")

Remove colunas não utilizadas e junta as informações com base nas colunas: `caso`, `tenure`, `max_iter` e `sol_ini`, para depois analisar cada caso de acordo com a média da configuração dos parâmetros.

In [5]:
test2_dropped = raw_test2_data.drop(columns=['num_teste'])

In [6]:
df = test2_dropped.groupby(['caso', 'tenure', 'max_iter', 'sol_ini']).agg({'custo': ['mean', 'min', 'max', np.std], 'tempo': ['mean', 'min', 'max', np.std], 'gap': ['mean', 'min', 'max', np.std]}).reset_index()

In [7]:
def fix_columns_name(df):
    return [column[0] if column[0] in ['caso', 'tenure', 'max_iter', 'sol_ini'] else column[0] + "_" + column[1] for column in df.columns]

In [8]:
df.columns = fix_columns_name(df)
df

caso  tenure  max_iter  sol_ini  custo_mean  custo_min  custo_max  \
0   A-n32-k5.vrp       5        10   Random      1055.6      940.0     1117.0   
1   A-n32-k5.vrp       5        10  Savings       887.8      850.0      913.0   
2   A-n32-k5.vrp       5       100   Random       955.6      903.0     1072.0   
3   A-n32-k5.vrp       5       100  Savings       846.6      835.0      857.0   
4   A-n32-k5.vrp       5       500   Random      1013.0      888.0     1117.0   
5   A-n32-k5.vrp       5       500  Savings       845.4      835.0      855.0   
6   A-n32-k5.vrp       5      1000   Random      1009.0      884.0     1073.0   
7   A-n32-k5.vrp       5      1000  Savings       839.6      835.0      850.0   
8   A-n32-k5.vrp      10        10   Random      1066.0      964.0     1115.0   
9   A-n32-k5.vrp      10        10  Savings       851.6      843.0      855.0   
10  A-n32-k5.vrp      10       100   Random      1070.6      989.0     1136.0   
11  A-n32-k5.vrp      10       100  Savings       857.8      835.0      913.0   
12  A-n32-k5.vrp      10       500   Random      1020.4      887.0     1124.0   
13  A-n32-k5.vrp      10       500  Savings       844.0      835.0      850.0   
14  A-n32-k5.vrp      10      1000   Random      1006.6      916.0     1094.0   
15  A-n32-k5.vrp      10      1000  Savings       835.0      835.0      835.0   
16  A-n32-k5.vrp      15        10   Random       983.8      913.0     1052.0   
17  A-n32-k5.vrp      15        10  Savings       867.6      855.0      913.0   
18  A-n32-k5.vrp      15       100   Random      1054.0      984.0     1143.0   
19  A-n32-k5.vrp      15       100  Savings       857.8      835.0      913.0   
20  A-n32-k5.vrp      15       500   Random      1027.0      859.0     1135.0   
21  A-n32-k5.vrp      15       500  Savings       838.0      835.0      850.0   
22  A-n32-k5.vrp      15      1000   Random      1034.8      872.0     1169.0   
23  A-n32-k5.vrp      15      1000  Savings       855.6      837.0      913.0   
24  A-n32-k5.vrp      16        10   Random      1000.2      965.0     1091.0   
25  A-n32-k5.vrp      16        10  Savings       850.0      845.0      855.0   
26  A-n32-k5.vrp      16       100   Random      1001.8      974.0     1018.0   
27  A-n32-k5.vrp      16       100  Savings       854.8      838.0      913.0   
28  A-n32-k5.vrp      16       500   Random      1014.0      968.0     1066.0   
29  A-n32-k5.vrp      16       500  Savings       845.2      835.0      853.0   
30  A-n32-k5.vrp      16      1000   Random       999.0      932.0     1072.0   
31  A-n32-k5.vrp      16      1000  Savings       853.6      835.0      913.0   
32  A-n32-k5.vrp      25        10   Random      1060.0      997.0     1131.0   
33  A-n32-k5.vrp      25        10  Savings       855.4      855.0      857.0   
34  A-n32-k5.vrp      25       100   Random      1000.6      926.0     1069.0   
35  A-n32-k5.vrp      25       100  Savings       852.8      835.0      913.0   
36  A-n32-k5.vrp      25       500   Random      1031.0      900.0     1117.0   
37  A-n32-k5.vrp      25       500  Savings       839.6      835.0      850.0   
38  A-n32-k5.vrp      25      1000   Random      1019.4      932.0     1139.0   
39  A-n32-k5.vrp      25      1000  Savings       867.8      835.0      913.0   
40  A-n54-k7.vrp       5        10   Random      1460.2     1352.0     1556.0   
41  A-n54-k7.vrp       5        10  Savings      1272.8     1265.0     1278.0   
42  A-n54-k7.vrp       5       100   Random      1504.4     1412.0     1575.0   
43  A-n54-k7.vrp       5       100  Savings      1272.8     1265.0     1278.0   
44  A-n54-k7.vrp       5       500   Random      1499.2     1412.0     1544.0   
45  A-n54-k7.vrp       5       500  Savings      1265.0     1265.0     1265.0   
46  A-n54-k7.vrp       5      1000   Random      1471.6     1435.0     1517.0   
47  A-n54-k7.vrp       5      1000  Savings      1267.6     1265.0     1278.0   
48  A-n54-k7.vrp      10        10   Random

## Casos

In [13]:
test2_dropped.sort_values(['gap', 'tempo']).head(10)

caso  tenure  max_iter  sol_ini  iter_sem_mod      tempo   custo  \
280  A-n54-k7.vrp      27        10  Savings            10   3.765352  1220.0   
48   A-n32-k5.vrp      15       100  Savings           100   3.436998   835.0   
65   A-n32-k5.vrp      25       100  Savings           100   3.522999   835.0   
27   A-n32-k5.vrp      10       100  Savings           100   3.688003   835.0   
8    A-n32-k5.vrp       5       100  Savings           100   4.516008   835.0   
70   A-n32-k5.vrp      25       500  Savings           500  13.973998   835.0   
30   A-n32-k5.vrp      10       500  Savings           500  14.019619   835.0   
90   A-n32-k5.vrp      16       500  Savings           500  14.046994   835.0   
34   A-n32-k5.vrp      10       500  Savings           500  14.114096   835.0   
51   A-n32-k5.vrp      15       500  Savings           500  14.167083   835.0   

          gap  
280  0.045416  
48   0.065051  
65   0.065051  
27   0.065051  
8    0.065051  
70   0.065051  
30   0.065051  
90   0.065051  
34   0.065051  
51   0.065051

In [14]:
df_a_n32_k5 = df[0:40]
df_a_n54_k7 = df[40:]

In [15]:
df_a_n32_k5.sort_values(['gap_mean', 'tempo_mean']).head(10)

caso  tenure  max_iter  sol_ini  custo_mean  custo_min  custo_max  \
15  A-n32-k5.vrp      10      1000  Savings       835.0      835.0      835.0   
21  A-n32-k5.vrp      15       500  Savings       838.0      835.0      850.0   
7   A-n32-k5.vrp       5      1000  Savings       839.6      835.0      850.0   
37  A-n32-k5.vrp      25       500  Savings       839.6      835.0      850.0   
13  A-n32-k5.vrp      10       500  Savings       844.0      835.0      850.0   
29  A-n32-k5.vrp      16       500  Savings       845.2      835.0      853.0   
5   A-n32-k5.vrp       5       500  Savings       845.4      835.0      855.0   
3   A-n32-k5.vrp       5       100  Savings       846.6      835.0      857.0   
25  A-n32-k5.vrp      16        10  Savings       850.0      845.0      855.0   
9   A-n32-k5.vrp      10        10  Savings       851.6      843.0      855.0   

    custo_std  tempo_mean  tempo_min  tempo_max  tempo_std  gap_mean  \
15   0.000000   27.807154  27.197988  28.631994   0.641695  0.065051   
21   6.708204   14.313636  13.955135  14.653004   0.264054  0.068878   
7    6.767570   30.560437  26.432996  41.672406   6.331228  0.070918   
37   6.308724   14.084634  13.532996  14.483524   0.390257  0.070918   
13   8.215838   13.811763  13.452014  14.114096   0.264986  0.076531   
29   9.391486   13.817950  13.317998  14.277072   0.375712  0.078061   
5    8.848729   17.836532  13.715002  25.946494   5.731113  0.078316   
3    9.208692    3.348798   2.937994   4.516008   0.656707  0.079847   
25   3.535534    0.619796   0.596997   0.644998   0.021041  0.084184   
9    5.272571    0.579199   0.571997   0.587004   0.005720  0.086224   

     gap_min   gap_max   gap_std  
15  0.065051  0.065051  0.000000  
21  0.065051  0.084184  0.008557  
7   0.065051  0.084184  0.008632  
37  0.065051  0.084184  0.008047  
13  0.065051  0.084184  0.010480  
29  0.065051  0.088010  0.011979  
5   0.065051  0.090561  0.011287  
3   0.065051  0.093112  0.011746  
25  0.077806  0.090561  0.004510  
9   0.075255  0.090561  0.006725

In [11]:
df_a_n54_k7.sort_values(['custo_mean', 'tempo_mean']).head(10)

caso  tenure  max_iter  sol_ini  custo_mean  custo_min  custo_max  \
73  A-n54-k7.vrp      27        10  Savings      1261.2     1220.0     1278.0   
49  A-n54-k7.vrp      10        10  Savings      1265.0     1265.0     1265.0   
57  A-n54-k7.vrp      15        10  Savings      1265.0     1265.0     1265.0   
77  A-n54-k7.vrp      27       500  Savings      1265.0     1265.0     1265.0   
53  A-n54-k7.vrp      10       500  Savings      1265.0     1265.0     1265.0   
45  A-n54-k7.vrp       5       500  Savings      1265.0     1265.0     1265.0   
59  A-n54-k7.vrp      15       100  Savings      1267.6     1265.0     1278.0   
51  A-n54-k7.vrp      10       100  Savings      1267.6     1265.0     1278.0   
75  A-n54-k7.vrp      27       100  Savings      1267.6     1265.0     1278.0   
61  A-n54-k7.vrp      15       500  Savings      1267.6     1265.0     1278.0   

    custo_std  tempo_mean  tempo_min  tempo_max  tempo_std  gap_mean  \
73  23.931151    3.341147   3.045544   3.765352   0.292479  0.080720   
49   0.000000    3.134407   3.036999   3.261018   0.084606  0.083976   
57   0.000000    3.138010   3.052067   3.188998   0.067308  0.083976   
77   0.000000   56.245306  55.370921  57.754325   0.922292  0.083976   
53   0.000000   57.168790  55.071102  64.189816   3.932770  0.083976   
45   0.000000   69.631079  62.249326  74.485362   4.591755  0.083976   
59   5.813777   12.564019  12.384997  12.702998   0.140196  0.086204   
51   5.813777   12.614222  12.406715  12.867431   0.193001  0.086204   
75   5.813777   13.170442  12.942745  13.410466   0.186445  0.086204   
61   5.813777   54.204658  53.326085  54.648738   0.515393  0.086204   

     gap_min   gap_max   gap_std  
73  0.045416  0.095116  0.020506  
49  0.083976  0.083976  0.000000  
57  0.083976  0.083976  0.000000  
77  0.083976  0.083976  0.000000  
53  0.083976  0.083976  0.000000  
45  0.083976  0.083976  0.000000  
59  0.083976  0.095116  0.004982  
51  0.083976  0.095116  0.004982  
75  0.083976  0.095116  0.004982  
61  0.083976  0.095116  0.004982

### A-n32-k5

A melhor combinação foi de:

`tenure = 10` <br>
`max_iter = 1000` <br>
`sol_ini= Savings` 

Porém, diversas combinações chegaram no custo mínimo de 835. Tem uma tendência de mais iterações apresentarem melhor resultado, porém a diferença média acaba sendo muito pequena e não justifica o ganho ao se compara com o tempo que dobra. Enquanto o caso A-n32-k5 seja pequeno, em uma situação em que o caso for maior 1000 iterações não apresenta ganho suficiente comparado a 500. Além disso, ao analisar o desvio ocorrido entre os experimentos, ele se mostra abaixo de 0.01 para alguns casos de iterações < 1000. 

In [106]:
df_a_n32_k5.sort_values(['custo_mean', 'tempo_mean']).head(10)

caso  tenure  max_iter  sol_ini  custo_mean  custo_min  custo_max  \
15  A-n32-k5.vrp      10      1000  Savings       835.0      835.0      835.0   
21  A-n32-k5.vrp      15       500  Savings       838.0      835.0      850.0   
37  A-n32-k5.vrp      25       500  Savings       839.6      835.0      850.0   
7   A-n32-k5.vrp       5      1000  Savings       839.6      835.0      850.0   
13  A-n32-k5.vrp      10       500  Savings       844.0      835.0      850.0   
29  A-n32-k5.vrp      16       500  Savings       845.2      835.0      853.0   
5   A-n32-k5.vrp       5       500  Savings       845.4      835.0      855.0   
3   A-n32-k5.vrp       5       100  Savings       846.6      835.0      857.0   
25  A-n32-k5.vrp      16        10  Savings       850.0      845.0      855.0   
9   A-n32-k5.vrp      10        10  Savings       851.6      843.0      855.0   

    custo_std  tempo_mean  tempo_min  tempo_max  tempo_std  gap_mean  \
15   0.000000   27.807154  27.197988  28.631994   0.641695  0.065051   
21   6.708204   14.313636  13.955135  14.653004   0.264054  0.068878   
37   6.308724   14.084634  13.532996  14.483524   0.390257  0.070918   
7    6.767570   30.560437  26.432996  41.672406   6.331228  0.070918   
13   8.215838   13.811763  13.452014  14.114096   0.264986  0.076531   
29   9.391486   13.817950  13.317998  14.277072   0.375712  0.078061   
5    8.848729   17.836532  13.715002  25.946494   5.731113  0.078316   
3    9.208692    3.348798   2.937994   4.516008   0.656707  0.079847   
25   3.535534    0.619796   0.596997   0.644998   0.021041  0.084184   
9    5.272571    0.579199   0.571997   0.587004   0.005720  0.086224   

     gap_min   gap_max   gap_std  
15  0.065051  0.065051  0.000000  
21  0.065051  0.084184  0.008557  
37  0.065051  0.084184  0.008047  
7   0.065051  0.084184  0.008632  
13  0.065051  0.084184  0.010480  
29  0.065051  0.088010  0.011979  
5   0.065051  0.090561  0.011287  
3   0.065051  0.093112  0.011746  
25  0.077806  0.090561  0.004510  
9   0.075255  0.090561  0.006725

#### Tenure = 10

Os melhores custos foram das soluções que tiveram como solução inicial a de savings. Em relação ao número máximo de iteração, o problema com 1000 max iterações parece apresentar o melhor custo, sendo de todos os testes o mínimo e máximo igual a 835. Porém, para testes com menos iterações também foi possível chegar nesse valor, como exemplo max_iter = 500 ou max_iter = 100.

A média com random apresenta um acrésimo de custo de mais que 160 em algumas simulações comparado ao melhor . 

In [131]:
df_a_n32_k5_t10 = df_a_n32_k5[df_a_n32_k5.tenure == 10]
df_a_n32_k5_t10.sort_values(['custo_mean', 'tempo_mean'])

caso  tenure  max_iter  sol_ini  custo_mean  custo_min  custo_max  \
15  A-n32-k5.vrp      10      1000  Savings       835.0      835.0      835.0   
13  A-n32-k5.vrp      10       500  Savings       844.0      835.0      850.0   
9   A-n32-k5.vrp      10        10  Savings       851.6      843.0      855.0   
11  A-n32-k5.vrp      10       100  Savings       857.8      835.0      913.0   
14  A-n32-k5.vrp      10      1000   Random      1006.6      916.0     1094.0   
12  A-n32-k5.vrp      10       500   Random      1020.4      887.0     1124.0   
8   A-n32-k5.vrp      10        10   Random      1066.0      964.0     1115.0   
10  A-n32-k5.vrp      10       100   Random      1070.6      989.0     1136.0   

    custo_std  tempo_mean  tempo_min  tempo_max  tempo_std  gap_mean  \
15   0.000000   27.807154  27.197988  28.631994   0.641695  0.065051   
13   8.215838   13.811763  13.452014  14.114096   0.264986  0.076531   
9    5.272571    0.579199   0.571997   0.587004   0.005720  0.086224   
11  31.507142    3.131601   2.913995   3.688003   0.324488  0.094133   
14  75.929573   34.118877  29.888627  37.350985   3.455947  0.283928   
12  93.160077   18.105222  15.037769  20.526976   2.113663  0.301531   
8   59.312731    1.075794   0.814086   1.259524   0.170011  0.359694   
10  52.628889    4.144555   3.889564   4.512805   0.262400  0.365561   

     gap_min   gap_max   gap_std  
15  0.065051  0.065051  0.000000  
13  0.065051  0.084184  0.010480  
9   0.075255  0.090561  0.006725  
11  0.065051  0.164541  0.040188  
14  0.168367  0.395408  0.096849  
12  0.131378  0.433673  0.118826  
8   0.229592  0.422194  0.075654  
10  0.261480  0.448980  0.067129

df_a_n32_k5_t10

In [140]:
df_a_n32_k5_t10_generic = test2_dropped.groupby(['caso', 'tenure', 'sol_ini']).agg({'custo': ['mean', 'min', 'max', np.std], 'tempo': ['mean', 'min', 'max', np.std], 'gap': ['mean', 'min', 'max', np.std]}).reset_index()
df_a_n32_k5_t10_generic.columns = fix_columns_name(df_a_n32_k5_t10_generic)

In [142]:
df_a_n32_k5_t10_generic = df_a_n32_k5_t10_generic.loc[(df_a_n32_k5_t10_generic.tenure == 10) & (df_a_n32_k5_t10_generic.caso == 'A-n32-k5.vrp')]
df_a_n32_k5_t10_generic.sort_values(['custo_mean', 'tempo_mean'])

caso  tenure  sol_ini  custo_mean  custo_min  custo_max  custo_std  \
3  A-n32-k5.vrp      10  Savings       847.1      835.0      913.0  17.480515   
2  A-n32-k5.vrp      10   Random      1040.9      887.0     1136.0  71.991154   

   tempo_mean  tempo_min  tempo_max  tempo_std  gap_mean   gap_min   gap_max  \
3   11.332429   0.571997  28.631994  11.013378  0.080485  0.065051  0.164541   
2   14.361112   0.814086  37.350985  13.557601  0.327679  0.131378  0.448980   

    gap_std  
3  0.022297  
2  0.091825

Ao analisar o tenure = 10, não parece ter uma diferença do número de iterações de forma impactante, sendo o desvio de 0.02 para Savings e 0.09 para Random.